src/work/Work.cpp

```diff
@@ -4,434 +4,224 @@
+                                  BasicWork::State::WORK_RUNNING;
+                       });
 }
 
-Work::State
-Work::onSuccess()
+bool
+Work::hasChildren() const
 {
-    return WORK_SUCCESS;
+    return !mChildren.empty();
 }
 
-void
-Work::onFailureRetry()
+bool
+Work::anyChildRaiseFailure() const
 {
+    return std::any_of(mChildren.begin(), mChildren.end(),
+                       [](std::shared_ptr<BasicWork> const& w) {
+                           return w->getState() ==
+                                  BasicWork::State::WORK_FAILURE;
+                       });
 }
 
-void
-Work::onFailureRaise()
+std::shared_ptr<BasicWork>
+Work::yieldNextRunningChild()
 {
-}
+    while (mNextChild != mChildren.end())
+    {
+        auto next = mNextChild;
+        mNextChild++;
+        assert(*next);
+        auto state = (*next)->getState();
+        if (state == State::WORK_RUNNING)
+        {
+            return *next;
+        }
+        else if (state == State::WORK_SUCCESS)
+        {
+            mNextChild = mChildren.erase(next);
+        }
+    }
 
-Work::State
-Work::getState() const
-{
-    return mState;
+    mNextChild = mChildren.begin();
+
+    return nullptr;
 }
 
-bool
-Work::isDone() const
+WorkSequence::WorkSequence(Application& app, std::string name,
+                           std::vector<std::shared_ptr<BasicWork>> sequence,
+                           size_t maxRetries)
+    : Work(app, std::move(name), maxRetries)
+    , mSequenceOfWork(sequence)
+    , mNextInSequence(mSequenceOfWork.begin())
 {
-    return mState == WORK_SUCCESS || mState == WORK_FAILURE_RAISE ||
-           mState == WORK_FAILURE_FATAL;
 }
 
-void
-Work::setState(Work::State st)
+BasicWork::State
+WorkSequence::doWork()
 {
-    auto maxR = getMaxRetries();
-    if (st == WORK_FAILURE_RETRY && (mRetries >= maxR))
+    if (mNextInSequence == mSequenceOfWork.end())
     {
-        CLOG(WARNING, "Work")
-            << "Reached retry limit " << maxR << " for " << getUniqueName();
-        st = WORK_FAILURE_RAISE;
+        // Completed all the work
+        assert(!hasChildren());
+        return State::WORK_SUCCESS;
     }
 
-    if (st != mState)
+    auto w = *mNextInSequence;
+    assert(w);
+    if (!hasChildren())
     {
-        CLOG(DEBUG, "Work") << "work " << getUniqueName() << " : "
-                            << stateName(mState) << " -> " << stateName(st);
-        mState = st;
+        w->startWork(wakeUpCallback());
+        addChild(w);
+        return State::WORK_RUNNING;
+    }
+    else
+    {
+        auto state = w->getState();
+        if (state == State::WORK_SUCCESS)
+        {
+            clearChildren();
+            mNextInSequence++;
+            return State::WORK_RUNNING;
+        }
+        return state;
     }
 }
 
 void
-Work::notifyParent()
+WorkSequence::doReset()
 {
-    auto parent = mParent.lock();
-    if (parent)
-    {
-        parent->notify(getUniqueName());
-    }
+    assert(std::all_of(
```

Comment:
`WorkSequence` itself resets the work before actually adding it (by calling startWork), so if somehow reset is called on work in the middle of running in the sequence, it would be a bug.